### DISCUSSION TITLE
Mod for failed upgrades/installations: AppID-notinstalled.ps1

### DISCUSSION THREAD

**Discussion by KnifMelti** [Romanitho/Winget-AutoUpdate#497] [MODs]
If something fails and we are left with an uninstalled application despite all the possibilities via already existing mods (in a SCCM environment)...

In `Update-App.ps1` just before  `Write-ToLog "##########   WINGET UPGRADE PROCESS FINISHED FOR APPLICATION ID '$($App.Id)'   ##########" "Gray"` we could have:
```
if ($FailedToUpgrade -eq $true) {
    if ($ModsNotInstalled) {
        Write-ToLog "Modifications for $($app.Id) after a failed upgrade/install are being applied..." "Yellow"
        & "$ModsNotInstalled"
    }
}
```

In that way one could for example, in this case, in a `Microsoft.SQLServerManagementStudio-notinstalled.ps1` delete a detection method and trigger SCCM to install the application anew:
```
$ConfirmInstall = Confirm-Installation $("Microsoft.SQLServerManagementStudio") $("19.")

if ($ConfirmInstall -ne $true) {
	$DelFile = @("${env:ProgramData}\Ssms19.fix")
	Remove-ModsFile $DelFile

	$computerName = $env:COMPUTERNAME

	try {
	    $client = [wmiclass] "\\$computerName\root\ccm:SMS_Client"
	    $Result = $client.TriggerSchedule("{00000000-0000-0000-0000-000000000121}")
	    Write-ToLog "Succesfully triggered a software updates schedule on $computerName" "Green"
	}
	catch {
	    Write-ToLog "Failed to trigger a software updates schedule on $computerName, Error: $_" "Red"
	}

}

```

**Comment by Romanitho** [comment]
And what about a double check?
- Check if update succeded
- Check if application is still installed if update failed. This case would enforce the app installation.

We can keep your mod implementation in case of specific action. But we probably need a generic action. I encounter this use case too with DotPaint.net

**Reply by KnifMelti** [reply]
But, we already have that...
```
    #Check if application updated properly
    $FailedToUpgrade = $false
    $ConfirmInstall = Confirm-Installation $($app.Id) $($app.AvailableVersion)

    if ($ConfirmInstall -ne $true) {
        #Upgrade failed!
        #Test for a Pending Reboot (Component Based Servicing/WindowsUpdate/CCM_ClientUtilities)
        $PendingReboot = Test-PendingReboot
        if ($PendingReboot -eq $true) {
            Write-ToLog "-> A Pending Reboot lingers and probably prohibited $($app.Name) from upgrading...`n-> ...an install for $($app.Name) is NOT executed!" "Red"
            $FailedToUpgrade = $true
            break
        }

        #If app failed to upgrade, run Install command
        Write-ToLog "-> An upgrade for $($app.Name) failed, now trying an install instead..." "Yellow"

```

**Reply by KnifMelti** [reply]
#498

**Reply by Romanitho** [reply]
Not exactly. Sometimes, Winget uninstalls the app (that may behave correctly) but fails on installing the update. Then, no application is installed anymore, and WAU says the update failed... On next run, WAU can't see the app (because previously uninstalled) and do not propose any update.
As a result, WAU uninstalled the app :(


So, here:
https://github.com/Romanitho/Winget-AutoUpdate/blob/17b02dfc9598af30f4b39e12bf11eb163e3cbc09/Winget-AutoUpdate/functions/Confirm-Installation.ps1#L19
we should have a check for the version (and tell if the app correctly updated) and another check to see if the app is still present. If not, then just re-install.

Hope i'm clear enough 😅

**Comment by KnifMelti** [comment]
> I encounter this use case too with DotPaint.net

In this example the need for extra checks comes from:

- Microsoft.SQLServerManagementStudio-preinstall:
Wait for "Ssms","azuredatastudio" if it's running
Uninstall all versions of "Microsoft SQL Server Management Studio*","Azure Data Studio*"

This is done because we don't want to end up with a multitude of different versions of **Microsoft SQL Server Management Studio** and in that process of uninstalling everything in beforehand it could end up as not installed alltogether (net drops etc...)
Hence the only detection mode in SCCM is "${env:ProgramData}\Ssms19.fix" and that file must be handled too.

So, as a fallback we trigger the SCCM agent to run a software update cycle so if it's not installed the agent will install the version from SCCM

**Reply by Romanitho** [reply]
Regarding your example with Management Studio, Wouldn't this be possible with existing Mods?

**Reply by KnifMelti** [reply]
Yes, of course we can do it all in the -preinstall.ps1

- Uninstall @("Microsoft SQL Server Management Studio*","Azure Data Studio*") via the native uninstall methods in registry
- Install `Microsoft.SQLServerManagementStudio` via **WinGet** (then it gets the newest version)
- If the WinGet installation fails (network error etc.) trigger **SCCM** to run a software update cycle (but then it will be an older version and it won't be installed until some time after **WAU** has finished)
- Let **WinGet** detect that the version already is installed (or needs an upgrade if it failed the installation in the -preinstall.ps1)
- Let **WAU** detect the correct version is installed or if it failed again mark it as failed.

But, much better to have an all-purpose mod (or an AppID mod for failed installations: WAU-notinstalled.ps1/AppID-notinstalled.ps1) that will be triggered **after** **WAU** detects a failure in my mind!

**Reply by Romanitho** [reply]
yes, it can be another mod. A real Swiss knife :)
